### PR TITLE
[dotnetup] Support tar.gz archives on Windows to enable hard link deduplication

### LINKED ISSUE #54403: dotnetup: Support tar.gz archives on Windows to enable hard link deduplication
## Summary

`dotnetup` on Windows currently only downloads and extracts `.zip` archives. This means Windows users installing via `dotnetup` cannot benefit from the hard link-based deduplication work in #52182, which relies on tar archives to preserve hard links.

## Background

The SDK deduplication effort (#52182) achieves ~28% size reduction on disk by using hard links to eliminate duplicate files. Tar archives natively preserve hard links, so Linux/macOS users get this benefit automatically. However, the `.zip` format does not support hard links — extracting a zip results in fully materialized duplicate files.

## Current Behavior

In `DotnetupUtilities.GetArchiveFileExtensionForPlatform()`:
```csharp
if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
    return ".zip";
else
    return ".tar.gz";
```

And in `DotnetArchiveExtractor.ExtractArchiveDirectlyToTarget()`:
```csharp
if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
    ExtractTarArchive(...);
else
    ExtractZipArchive(...);
```

The format decision is purely OS-based with no fallback or user option.

## Proposed Change

Support downloading and extracting `.tar.gz` archives on Windows when available. This would:

1. Allow Windows users to benefit from the ~28% on-disk size reduction from hard link deduplication
2. Reduce download size (tar.gz is smaller than zip for SDK payloads — 208 MB vs 289 MB per #52182)
3. Preserve hard links during extraction using `System.Formats.Tar` (which already works cross-platform)

### Considerations

- NTFS supports hard links, so the extraction logic should work on Windows
- The existing `ExtractTarArchive` code uses `System.Formats.Tar` which is cross-platform
- May want to prefer tar.gz when available but fall back to zip if the tar.gz artifact is not published for a given release
- Need to verify symlink/hardlink extraction works correctly with Windows filesystem semantics and permissions

## Related

- #52182 — Eliminate Duplicate SDK Files (hard link deduplication design)

### PR DESCRIPTION
**Pull request by MichaelSimons**
## Summary

Enables dotnetup to use tar.gz archives on Windows, which support hard links for disk space deduplication and are smaller in size, resulting in faster downloads. Previously, Windows was hardcoded to use .zip archives which don't support hard links.

Fixes #54403

## Changes

- **Prefer tar.gz on all platforms** — `ReleaseManifest.FindMatchingFile()` and `ResolveBlobFeedEntry()` now prefer `.tar.gz` archives with a `.zip` fallback on Windows for older .NET versions that may not publish tar.gz for Windows.
- **Route extraction by file extension, not OS** — `ExtractArchiveDirectlyToTarget()` checks the archive file extension to determine whether to use tar or zip extraction, rather than checking the OS platform.
- **Single manifest lookup** — `DownloadArchiveWithVerification()` now accepts a base path (without extension), resolves the archive format during download, appends the extension, and returns the full path. This eliminates a redundant manifest lookup that was previously needed to determine the file extension up front.
- **Remove `DotnetupUtilities.GetArchiveFileExtensionForPlatform()`** — no longer needed since the format is determined dynamically from the resolved download URL.
- **Improved error messages** — when blob feed resolution fails, the error message lists all checksum URLs that were probed.

## Test coverage

- Updated `ResolveManifestEntry_BlobFeed404_ThrowsVersionNotFound` to account for the tar.gz + zip probing on Windows.
- Added `ResolveManifestEntry_FallsBackToZip_WhenTarGzNotAvailable_OnWindows` — validates the zip fallback in blob feed resolution.
- Added `Commit_ExtractsTarGzArchive_Correctly` — end-to-end test creating a real tar.gz and verifying extraction.
- Added `Commit_ExtractsZipArchive_Correctly` (Windows-only) — end-to-end test creating a real zip and verifying extraction.

### CODE REVIEW

File: src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs [hunk 1]
```diff
@@ -175,29 +175,33 @@ private void DownloadArchive(string downloadUrl, string destinationPath, IProgre
...
-    /// Downloads the archive for the specified installation and verifies its hash.
-    /// Checks the download cache first to avoid re-downloading.
+    /// Downloads the archive for the specified installation request and verifies its hash.
+    /// The implementation resolves the appropriate archive format and appends the file extension
+    /// to the provided base path.
     /// </summary>
     /// <param name="installRequest">The .NET installation request details</param>
     /// <param name="resolvedVersion">The resolved version to download</param>
-    /// <param name="destinationPath">The local path to save the downloaded file</param>
+    /// <param name="destinationBasePath">The local base path (without extension) to save the downloaded file</param>
```
Comment (nagilson) [inline src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs:184]: Nice call out here that we dont want the file extension here.

File: src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs [hunk 2]
```diff
@@ -209,6 +213,22 @@ public void DownloadArchiveWithVerification(
...
+
+        return destinationPath;
+    }
+
+    /// <summary>
+    /// Extracts the archive file extension from a download URL.
+    /// </summary>
+    private static string GetExtensionFromUrl(string url)
+    {
+        if (url.EndsWith(".tar.gz", StringComparison.OrdinalIgnoreCase))
```
Comment (nagilson) [inline src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs]: Is this here to skip having to build a uri (perf optimization)? It might be helpful to add a comment below if so since the code technically works fine without this line I believe.

File: src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs [hunk 3]
```diff
@@ -315,18 +335,31 @@ private static bool ShouldFallbackToBlobFeed(
...
         string rid = DotnetupUtilities.GetRuntimeIdentifier(installRequest.InstallRoot.Architecture);
-        string extension = DotnetupUtilities.GetArchiveFileExtensionForPlatform();
+        var locationsChecked = new List<string>();
 
-        var location = BlobFeedUrlBuilder.GetFeedLocation(installRequest.Component, resolvedVersion, rid, extension);
-        string? hash = TryGetHashFromUrl(location.ChecksumUrl, resolvedVersion, installRequest.Component);
+        // Prefer tar.gz, fall back to zip if unavailable (older .NET versions may not publish tar.gz for Windows)
+        var tarLocation = BlobFeedUrlBuilder.GetFeedLocation(installRequest.Component, resolvedVersion, rid, ".tar.gz");
+        locationsChecked.Add(tarLocation.ChecksumUrl);
+        string? hash = TryGetHashFromUrl(tarLocation.ChecksumUrl, resolvedVersion, installRequest.Component);
```
Comment (nagilson) [inline src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs:347]: Part of me wants to make `TryGetHashFromUrl` add to `locationsChecked` but also that seems overkill

File: src/Installer/Microsoft.Dotnet.Installation/Internal/ReleaseManifest.cs [hunk 4]
```diff
@@ -190,15 +190,25 @@ internal static bool IsMatchingRuntimeComponent(string componentName, InstallCom
...
              .Where(f => !IsCompositeArchive(f.Name))
              .Where(f => !IsApphostPackArchive(f.Name))
              .ToList();
 
+        // Prefer tar.gz, fall back to zip if unavailable (older .NET versions may not publish tar.gz for Windows)
+        var matchingFiles = ridFiles
+             .Where(f => f.Name.EndsWith(".tar.gz", StringComparison.OrdinalIgnoreCase))
+             .ToList();
+
+        if (matchingFiles.Count == 0 && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (nagilson) [inline src/Installer/Microsoft.Dotnet.Installation/Internal/ReleaseManifest.cs]: nit: could do !matchingFiles.Any() ?

File: test/dotnetup.Tests/DnupE2Etest.cs [hunk 5]
```diff
@@ -387,7 +387,7 @@ public void SdkInstall_FullySpecifiedPreviewVersion_UsesBlobFeedFallback()
     private static string? TryResolveCurrentDailyPreviewSdkVersion()
     {
         string rid = DotnetupUtilities.GetRuntimeIdentifier(InstallerUtilities.GetDefaultInstallArchitecture());
-        string ext = DotnetupUtilities.GetArchiveFileExtensionForPlatform();
+        string ext = ".tar.gz";
```
Comment (nagilson) [inline test/dotnetup.Tests/DnupE2Etest.cs]: This is our own technical debt, but maybe we'd want to add a DefaultArchiveFileExtension constant for tests to consume at minimum or simply construct this in a fucntion so we dont need a rid and ext to add every time per test

File: test/dotnetup.Tests/DotnetArchiveExtractorTests.cs [hunk 6]
```diff
@@ -419,5 +419,129 @@ public void ExtractTarContents_RootLevelFilesAlwaysExtracted()
...
+
+        using var extractor = new DotnetArchiveExtractor(request, version, releaseManifest, progressTarget, mockDownloader);
+
+        // Act
+        extractor.Prepare();
+        extractor.Commit();
+
+        // Assert
+        var runtimeFile = Path.Combine(testEnv.InstallPath, "shared", "Microsoft.NETCore.App", "9.0.0", "System.Runtime.dll");
+        File.Exists(runtimeFile).Should().BeTrue("tar.gz extraction should create files in the install path");
```
Comment (nagilson) [inline test/dotnetup.Tests/DotnetArchiveExtractorTests.cs:460]: should we have any validation on the file permissions extracted on win here?

**Review by nagilson — APPROVED** [review]
Thanks for taking this up!

File: src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs (hunk 3 above)
Comment (MichaelSimons) [reply src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs:347]: I had intially created a helper method for the GetFeedLocation, locationsChecked and TryGetHashFromUrl but I felt it made the code more complicated/hurt the readability.  I therefore settled on this pattern that does have a slight bit of duplication.  Leaving as is unless someone convinces me otherwise/feels strongly about this.

File: src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs (hunk 2 above)
Comment (MichaelSimons) [reply src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveDownloader.cs]: It really should be using the uri.LocalPath if we don't want the method to be making assumptions about the url being passed.  I refactored the implementation in https://github.com/dotnet/sdk/pull/54467/commits/4a75516d6a7805c5b51ce8864bf47da7623d51c6

File: src/Installer/Microsoft.Dotnet.Installation/Internal/ReleaseManifest.cs (hunk 4 above)
Comment (MichaelSimons) [reply src/Installer/Microsoft.Dotnet.Installation/Internal/ReleaseManifest.cs]: fixed in https://github.com/dotnet/sdk/pull/54467/commits/4a75516d6a7805c5b51ce8864bf47da7623d51c6

File: test/dotnetup.Tests/DnupE2Etest.cs (hunk 5 above)
Comment (MichaelSimons) [reply test/dotnetup.Tests/DnupE2Etest.cs]: I added a constant in https://github.com/dotnet/sdk/pull/54467/commits/4a75516d6a7805c5b51ce8864bf47da7623d51c6.  I looked at a method but the use cases across the tests mostly feel different - this felt beyond the scope of this PR if even valuable.

File: test/dotnetup.Tests/DotnetArchiveExtractorTests.cs (hunk 6 above)
Comment (MichaelSimons) [reply test/dotnetup.Tests/DotnetArchiveExtractorTests.cs:460]: I didn't see any existing permissions tests for the zip so I didn't define any.  Regardless this seems of limited value.  I am going to leave as is unless you feel strongly about the value here.

File: test/dotnetup.Tests/DnupE2Etest.cs (hunk 5 above)
Comment (nagilson) [reply test/dotnetup.Tests/DnupE2Etest.cs]: Makes sense, thanks for documenting and considering my feedback.

**Review by nagilson — APPROVED** [review]
Thanks for thoughtfully addressing my feedback, happy with the changes. :shipit:
